presto-main/src/main/java/com/facebook/presto/spiller/BinaryFileSingleStreamSpillerFactory.java

```diff
@@ -0,0 +1,106 @@
+
+    private final ListeningExecutorService executor;
+    private final PagesSerdeFactory serdeFactory;
+    private final List<Path> spillPaths;
+    private final SpillerStats spillerStats;
+    private final double minimumFreeSpaceThreshold;
+    private int roundRobinIndex;
+
+    @Inject
+    public BinaryFileSingleStreamSpillerFactory(BlockEncodingSerde blockEncodingSerde, SpillerStats spillerStats, FeaturesConfig featuresConfig)
+    {
+        this(createExecutorServiceOfSize(requireNonNull(featuresConfig, "featuresConfig is null").getSpillerThreads()),
+                blockEncodingSerde,
+                spillerStats,
+                requireNonNull(featuresConfig, "featuresConfig is null").getSpillerSpillPaths(),
+                requireNonNull(featuresConfig, "featuresConfig is null").getSpillMaxUsedSpaceThreshold());
+    }
+
+    public BinaryFileSingleStreamSpillerFactory(
+            ListeningExecutorService executor,
+            BlockEncodingSerde blockEncodingSerde,
+            SpillerStats spillerStats,
+            List<Path> spillPaths,
+            double maxUsedSpaceThreshold)
+    {
+        this.serdeFactory = new PagesSerdeFactory(requireNonNull(blockEncodingSerde, "blockEncodingSerde is null"), false);
+        this.executor = requireNonNull(executor, "executor is null");
+        this.spillerStats = requireNonNull(spillerStats, "spillerStats can not be null");
+        requireNonNull(spillPaths, "spillPaths is null");
+        checkArgument(spillPaths.size() >= 1, "At least one spill path required");
+        this.spillPaths = ImmutableList.copyOf(spillPaths);
+        spillPaths.forEach(path -> path.toFile().mkdirs());
+        this.minimumFreeSpaceThreshold = requireNonNull(maxUsedSpaceThreshold, "maxUsedSpaceThreshold can not be null");
+        this.roundRobinIndex = 0;
+    }
+
+    private static ListeningExecutorService createExecutorServiceOfSize(int nThreads)
+    {
+        ThreadFactory threadFactory = daemonThreadsNamed(SPILLER_THREAD_NAME_PREFIX + "-%s");
+        ExecutorService executorService = newFixedThreadPool(nThreads, threadFactory);
+        return MoreExecutors.listeningDecorator(executorService);
+    }
+
+    @Override
+    public SingleStreamSpiller create(List<Type> types, LocalSpillContext localSpillContext)
+    {
+        return new BinaryFileSingleStreamSpiller(serdeFactory.createPagesSerde(), executor, getNextSpillPath(), spillerStats, localSpillContext);
+    }
+
+    private synchronized Path getNextSpillPath()
+    {
+        int spillPathsCount = spillPaths.size();
+        for (int i = 0; i < spillPathsCount; ++i) {
+            Path path = spillPaths.get((roundRobinIndex + i) % spillPathsCount);
+            if (hasEnoughDiskSpace(path)) {
+                roundRobinIndex = (roundRobinIndex + i + 1) % spillPathsCount;
+                return path;
+            }
+        }
+        throw new RuntimeException("No free space available for spill");
```

Comment:
Added `OUT_OF_SPILL_SPACE` error code in class `INTERNAL_ERROR` and changed this to `PrestoException`.